### PR TITLE
Camera modes

### PR DESCRIPTION
**Pull request by Frickerson**

### CODE REVIEW

File: Source/GameBaseFramework/Private/Abilities/GBFGameplayAbility.cpp [hunk 1]
```diff
@@ -0,0 +1,51 @@
+﻿#include "Abilities/GBFGameplayAbility.h"
+
+#include "Camera/GBFCameraMode.h"
+#include "Characters/Components/GBFHeroComponent.h"
+
+#include <AbilitySystemLog.h>
+
+#define ENSURE_ABILITY_IS_INSTANTIATED_OR_RETURN( FunctionName, ReturnValue )                                                                                \
+    {                                                                                                                                                        \
+        if ( !ensure( IsInstantiated() ) )                                                                                                                   \
```
Comment (TheEmidee) [inline Source/GameBaseFramework/Private/Abilities/GBFGameplayAbility.cpp]: ensureAlways

File: Source/GameBaseFramework/Private/Abilities/GBFGameplayAbility.cpp [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    }
+
+UGBFGameplayAbility::UGBFGameplayAbility()
+{
+    ActiveCameraMode = nullptr;
+}
+
+UGBFHeroComponent * UGBFGameplayAbility::GetHeroComponentFromActorInfo() const
+{
+    return ( CurrentActorInfo ? UGBFHeroComponent::FindHeroComponent( CurrentActorInfo->AvatarActor.Get() ) : nullptr );
```
Comment (TheEmidee) [inline Source/GameBaseFramework/Private/Abilities/GBFGameplayAbility.cpp]: you can remove the surrounding parenthesis

File: Source/GameBaseFramework/Private/Camera/GBFCameraComponent.cpp [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+}
+
+void UGBFCameraComponent::AddFieldOfViewOffset( const float fov_offset )
+{
+    FieldOfViewOffset += fov_offset;
+}
+
+void UGBFCameraComponent::GetBlendInfo( float & out_weight_of_top_layer, FGameplayTag & out_tag_of_top_layer ) const
+{
+    check( CameraModeStack );
```
Comment (TheEmidee) [inline Source/GameBaseFramework/Private/Camera/GBFCameraComponent.cpp]: explicit nullcheck

File: Source/GameBaseFramework/Private/Camera/GBFCameraComponent.cpp [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    CameraModeStack->GetBlendInfo( out_weight_of_top_layer, out_tag_of_top_layer );
+}
+
+void UGBFCameraComponent::OnRegister()
+{
+    Super::OnRegister();
+
+    if ( !CameraModeStack )
```
Comment (TheEmidee) [inline Source/GameBaseFramework/Private/Camera/GBFCameraComponent.cpp]: check against nullptr

File: Source/GameBaseFramework/Private/Camera/GBFCameraComponent.cpp [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+    if ( IsXRHeadTrackedCamera() )
+    {
+        // In XR much of the camera behavior above is irrellevant, but the post process settings are not.
+        Super::GetCameraView( delta_time, desired_view );
+    }
+}
+
+void UGBFCameraComponent::UpdateCameraModes()
+{
+    check( CameraModeStack );
```
Comment (TheEmidee) [inline Source/GameBaseFramework/Private/Camera/GBFCameraComponent.cpp]: != nullptr

File: Source/GameBaseFramework/Private/Camera/GBFCameraComponent.cpp [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+    if ( !CameraModeStack )
+    {
+        CameraModeStack = NewObject< UGBFCameraModeStack >( this );
+        check( CameraModeStack );
+    }
+}
+
+void UGBFCameraComponent::GetCameraView( const float delta_time, FMinimalViewInfo & desired_view )
+{
+    check( CameraModeStack );
```
Comment (TheEmidee) [inline Source/GameBaseFramework/Private/Camera/GBFCameraComponent.cpp]: same

File: Source/GameBaseFramework/Private/Camera/GBFCameraComponent.cpp [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+}
+
+void UGBFCameraComponent::OnRegister()
+{
+    Super::OnRegister();
+
+    if ( !CameraModeStack )
+    {
+        CameraModeStack = NewObject< UGBFCameraModeStack >( this );
+        check( CameraModeStack );
```
Comment (TheEmidee) [inline Source/GameBaseFramework/Private/Camera/GBFCameraComponent.cpp]: != nullptr

File: Source/GameBaseFramework/Private/Camera/GBFCameraMode.cpp [hunk 8]
```diff
@@ -0,0 +1,454 @@
...
+    const auto * target_actor = GetTargetActor();
+    check( target_actor );
+
+    if ( const auto * target_pawn = Cast< APawn >( target_actor ) )
+    {
+        // Height adjustments for characters to account for crouching.
+        if ( const auto * target_character = Cast< ACharacter >( target_pawn ) )
+        {
+            const auto * target_character_cdo = target_character->GetClass()->GetDefaultObject< ACharacter >();
+            check( target_character_cdo );
```
Comment (TheEmidee) [inline Source/GameBaseFramework/Private/Camera/GBFCameraMode.cpp]: != nullptr everywhere

File: Source/GameBaseFramework/Private/Camera/GBFCameraMode.cpp [hunk 9]
```diff
@@ -0,0 +1,454 @@
...
+        return target_pawn->GetPawnViewLocation();
+    }
+
+    return target_actor->GetActorLocation();
+}
+
+FRotator UGBFCameraMode::GetPivotRotation() const
+{
+    const auto * target_actor = GetTargetActor();
+    check( target_actor );
```
Comment (TheEmidee) [inline Source/GameBaseFramework/Private/Camera/GBFCameraMode.cpp]: here

File: Source/GameBaseFramework/Private/Camera/GBFCameraMode.cpp [hunk 10]
```diff
@@ -0,0 +1,454 @@
...
+void UGBFCameraMode::SetBlendWeight( const float weight )
+{
+    BlendWeight = FMath::Clamp( weight, 0.0f, 1.0f );
+
+    // Since we're setting the blend weight directly, we need to calculate the blend alpha to account for the blend function.
+    const auto inv_exponent = ( BlendExponent > 0.0f ) ? ( 1.0f / BlendExponent ) : 1.0f;
+
+    switch ( BlendFunction )
+    {
+        case EGBFCameraModeBlendFunction::Linear:
```
Comment (TheEmidee) [inline Source/GameBaseFramework/Private/Camera/GBFCameraMode.cpp:158]: missing {} in the case statements

File: Source/GameBaseFramework/Private/Camera/GBFCameraMode.cpp [hunk 11]
```diff
@@ -0,0 +1,454 @@
...
+    {
+        BlendAlpha = 1.0f;
+    }
+
+    const auto exponent = ( BlendExponent > 0.0f ) ? BlendExponent : 1.0f;
+
+    switch ( BlendFunction )
+    {
+        case EGBFCameraModeBlendFunction::Linear:
+            BlendWeight = BlendAlpha;
```
Comment (TheEmidee) [inline Source/GameBaseFramework/Private/Camera/GBFCameraMode.cpp:204]: same

File: Source/GameBaseFramework/Private/Camera/GBFCameraMode.cpp [hunk 12]
```diff
@@ -0,0 +1,454 @@
...
+void UGBFCameraModeStack::ActivateStack()
+{
+    if ( !bIsActive )
+    {
+        bIsActive = true;
+
+        // Notify camera modes that they are being activated.
+        for ( auto & camera_mode : CameraModeStack )
+        {
+            check( camera_mode );
```
Comment (TheEmidee) [inline Source/GameBaseFramework/Private/Camera/GBFCameraMode.cpp]: != nullptr

File: Source/GameBaseFramework/Private/Camera/GBFCameraMode.cpp [hunk 13]
```diff
@@ -0,0 +1,454 @@
...
+void UGBFCameraModeStack::DeactivateStack()
+{
+    if ( bIsActive )
+    {
+        bIsActive = false;
+
+        // Notify camera modes that they are being deactivated.
+        for ( auto & camera_mode : CameraModeStack )
+        {
+            check( camera_mode );
```
Comment (TheEmidee) [inline Source/GameBaseFramework/Private/Camera/GBFCameraMode.cpp]: same

File: Source/GameBaseFramework/Private/Camera/GBFCameraMode.cpp [hunk 14]
```diff
@@ -0,0 +1,454 @@
...
+            camera_mode->OnDeactivation();
+        }
+    }
+}
+
+void UGBFCameraModeStack::PushCameraMode( const TSubclassOf< UGBFCameraMode > camera_mode_class )
+{
+    if ( !camera_mode_class )
```
Comment (TheEmidee) [inline Source/GameBaseFramework/Private/Camera/GBFCameraMode.cpp]: same

File: Source/GameBaseFramework/Private/Camera/GBFCameraMode.cpp [hunk 15]
```diff
@@ -0,0 +1,454 @@
...
+    {
+        return;
+    }
+
+    auto * camera_mode = GetCameraModeInstance( camera_mode_class );
+    check( camera_mode );
```
Comment (TheEmidee) [inline Source/GameBaseFramework/Private/Camera/GBFCameraMode.cpp]: here

File: Source/GameBaseFramework/Private/Camera/GBFCameraMode.cpp [hunk 16]
```diff
@@ -0,0 +1,454 @@
...
+
+    auto stack_size = CameraModeStack.Num();
+
+    if ( ( stack_size > 0 ) && ( CameraModeStack[ 0 ] == camera_mode ) )
```
Comment (TheEmidee) [inline Source/GameBaseFramework/Private/Camera/GBFCameraMode.cpp]: unneeded parenthesis

File: Source/GameBaseFramework/Private/Camera/GBFCameraMode.cpp [hunk 17]
```diff
@@ -0,0 +1,454 @@
...
+
+    UpdateStack( delta_time );
+    BlendStack( out_camera_mode_view );
+
+    return true;
+}
+
+UGBFCameraMode * UGBFCameraModeStack::GetCameraModeInstance( const TSubclassOf< UGBFCameraMode > camera_mode_class )
+{
+    check( camera_mode_class );
```
Comment (TheEmidee) [inline Source/GameBaseFramework/Private/Camera/GBFCameraMode.cpp]: here

File: Source/GameBaseFramework/Private/Camera/GBFCameraMode.cpp [hunk 18]
```diff
@@ -0,0 +1,454 @@
...
+
+    // First see if we already created one.
+    for ( auto & camera_mode : CameraModeInstances )
+    {
+        if ( ( camera_mode != nullptr ) && ( camera_mode->GetClass() == camera_mode_class ) )
```
Comment (TheEmidee) [inline Source/GameBaseFramework/Private/Camera/GBFCameraMode.cpp]: extra ()

File: Source/GameBaseFramework/Private/Camera/GBFCameraMode.cpp [hunk 19]
```diff
@@ -0,0 +1,454 @@
...
+        {
+            return camera_mode.Get();
+        }
+    }
+
+    // Not found, so we need to create it.
+    auto * new_camera_mode = NewObject< UGBFCameraMode >( GetOuter(), camera_mode_class, NAME_None, RF_NoFlags );
+    check( new_camera_mode );
```
Comment (TheEmidee) [inline Source/GameBaseFramework/Private/Camera/GBFCameraMode.cpp]: here

File: Source/GameBaseFramework/Private/Characters/Components/GBFHeroComponent.cpp [hunk 20]
```diff
@@ -16,13 +18,32 @@ UGBFHeroComponent::UGBFHeroComponent()
...
 }
 
 UGBFHeroComponent * UGBFHeroComponent::FindHeroComponent( const AActor * actor )
 {
     return actor ? actor->FindComponentByClass< UGBFHeroComponent >() : nullptr;
 }
 
+void UGBFHeroComponent::SetAbilityCameraMode( TSubclassOf< UGBFCameraMode > camera_mode, const FGameplayAbilitySpecHandle & owning_spec_handle )
+{
+    if ( camera_mode )
```
Comment (TheEmidee) [inline Source/GameBaseFramework/Private/Characters/Components/GBFHeroComponent.cpp]: != nullptr

File: Source/GameBaseFramework/Private/Characters/Components/GBFHeroComponent.cpp [hunk 21]
```diff
@@ -174,3 +200,27 @@ void UGBFHeroComponent::OnPawnReadyToInitialize()
 void UGBFHeroComponent::InitializePlayerInput( UInputComponent * player_input_component )
 {
 }
+
+TSubclassOf< UGBFCameraMode > UGBFHeroComponent::DetermineCameraMode() const
+{
+    if ( AbilityCameraMode )
```
Comment (TheEmidee) [inline Source/GameBaseFramework/Private/Characters/Components/GBFHeroComponent.cpp]: != nullptr

File: Source/GameBaseFramework/Private/Characters/Components/GBFHeroComponent.cpp [hunk 22]
```diff
@@ -174,3 +200,27 @@ void UGBFHeroComponent::OnPawnReadyToInitialize()
...
+
+TSubclassOf< UGBFCameraMode > UGBFHeroComponent::DetermineCameraMode() const
+{
+    if ( AbilityCameraMode )
+    {
+        return AbilityCameraMode;
+    }
+
+    const auto * pawn = GetPawn< APawn >();
+    if ( !pawn )
```
Comment (TheEmidee) [inline Source/GameBaseFramework/Private/Characters/Components/GBFHeroComponent.cpp]: here

File: Source/GameBaseFramework/Public/Camera/GBFCameraMode.h [hunk 23]
```diff
@@ -0,0 +1,203 @@
...
+
+    const FGBFCameraModeView & GetCameraModeView() const;
+    float GetBlendTime() const;
+    float GetBlendWeight() const;
+    FGameplayTag GetCameraTypeTag() const;
+
+    UWorld * GetWorld() const override;
+
+    // Called when this camera mode is activated on the camera mode stack.
+    virtual void OnActivation() {};
```
Comment (TheEmidee) [inline Source/GameBaseFramework/Public/Camera/GBFCameraMode.h]: empty implementation are in the CPP file too

**Review by TheEmidee — CHANGES_REQUESTED** [review]

File: Source/GameBaseFramework/Public/Camera/GBFCameraMode.h [hunk 24]
```diff
@@ -0,0 +1,203 @@
...
+#include <CoreMinimal.h>
+#include <GameplayTagContainer.h>
+#include <UObject/Object.h>
+
+#include "GBFCameraMode.generated.h"
+
+class UGBFCameraComponent;
+class UCanvas;
+
+#define GBF_CAMERA_DEFAULT_FOV ( 80.0f )
```
Comment (TheEmidee) [inline Source/GameBaseFramework/Public/Camera/GBFCameraMode.h]: urk

I would replace with something like

```
namespace GBF
{
    namespace CameraMode
{
    static const float DefaultFov;
// etc...
}
}
```

and the matching initialization in the cpp file

**Review by TheEmidee — CHANGES_REQUESTED** [review]

File: Source/GameBaseFramework/Public/Camera/GBFCameraMode.h (hunk 24 above)
Comment (Frickerson) [reply Source/GameBaseFramework/Public/Camera/GBFCameraMode.h]: No need for the C++ init, if we want them const we need to init in header

**Review by TheEmidee — APPROVED** [review]

**Pull request merged**